pkg/upgrade/job/job.go

```diff
@@ -211,6 +214,21 @@ func New(plan *upgradeapiv1.Plan, node *corev1.Node, controllerName string) *bat
 		*job.Spec.Parallelism = 1
 	}
 
+	if plan.Spec.Exclusive {
+		job.Spec.Template.Spec.Affinity.PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution = []corev1.PodAffinityTerm{{
```

Comment:
Curious if we want to replace the default PodAntiAffinity expressions, or just append to them? Appending to them might be more intuitive, and require less boilerplate?

Reply:
I tried that initially, it did not work because the scheduler evaluates multiple anti-affinity rules with a logical OR when making scheduling decisions, so effectively the current scheduling would always happen regardless of additional rules.

Reply:
Hmm, that's interesting - the docs say that they should be ANDed:
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#podaffinity-v1-core
> **requiredDuringSchedulingIgnoredDuringExecution**
> [PodAffinityTerm](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.27/#podaffinityterm-v1-core) array	If the affinity requirements specified by this field are not met at scheduling time, the pod will not be scheduled onto the node. If the affinity requirements specified by this field cease to be met at some point during pod execution (e.g. due to a pod label update), the system may or may not try to eventually evict the pod from its node. When there are multiple elements, the lists of nodes corresponding to each podAffinityTerm are intersected, i.e. all terms must be satisfied.

Specifically the bit about `all terms must be satisfied` makes it clear that this is logical AND not OR.

Reply:
That is very interesting, I did not see that part of the docs when I did my previous testing so I just tested this again and can confirm the docs are not correct, the terms are indeed being OR'd:

1. Made a new commit in my fork using multiple tags [here](https://github.com/rancher/system-upgrade-controller/blob/721d70c57e91036470bc1d2e894e0b4a141069c2/pkg/upgrade/job/job.go#L173)
2. Built & pushed image, deployed to my cluster, applied my test yaml with multiple exclusive Plans
3. Pods all have the correct labels and anti-affinity rules and are still being co-located following the current default scheduling pattern:
```
jrodonnell@k3s0:~$ kubectl get pod -A -o wide --field-selector spec.nodeName=k3s0
NAMESPACE            NAME                                                              READY   STATUS            RESTARTS       AGE    IP            NODE   NOMINATED NODE   READINESS GATES
...
system-upgrade       apply-exclusive-true-1-on-k3s0-with-55d6cc304f29efcbfd4ba-6sxr8   1/1     Running   0              14s    192.168.1.4   k3s0   <none>           <none>
system-upgrade       apply-exclusive-true-2-on-k3s0-with-55d6cc304f29efcbfd4ba-xrcfs   1/1     Running   0              14s    192.168.1.4   k3s0   <none>           <none>

jrodonnell@k3s0:~$ k get pod apply-exclusive-true-1-on-k3s0-with-55d6cc304f29efcbfd4ba-6sxr8 -o json -n system-upgrade | jq -r '.metadata.labels'                                                                                                                                  
{                                                                                                                                                                                                                                                                                  
  "controller-uid": "a62f8de5-88be-4d61-9ea8-4c3878fee9e6",                                                                                                                                                                                                                        
  "job-name": "apply-exclusive-true-1-on-k3s0-with-55d6cc304f29efcbfd4ba-069ef",                                                                                                                                                                                                   
  "plan.upgrade.cattle.io/exclusive-true-1": "55d6cc304f29efcbfd4ba2d6253edb4f91b638b2afe0ce7b99aa8d82",                                                                                                                                                                           
  "upgrade.cattle.io/controller": "system-upgrade-controller",                                                                                                                                                                                                                     
  "upgrade.cattle.io/exclusive": "true",                                                                                                                                                                                                                                           
  "upgrade.cattle.io/node": "k3s0",                                                                                                                                                                                                                                                
  "upgrade.cattle.io/plan": "exclusive-true-1",                                                                                                                                                                                                                                    
  "upgrade.cattle.io/version": "bionic"
}

jrodonnell@k3s0:~$ k get pod apply-exclusive-true-1-on-k3s0-with-55d6cc304f29efcbfd4ba-6sxr8 -o json -n system-upgrade | jq -r '.spec.affinity.podAntiAffinity'
      ...
        "matchExpressions": [
          {
            "key": "upgrade.cattle.io/plan",
            "operator": "In",
            "values": [
              "exclusive-true-1"
            ]
          },
          {
            "key": "upgrade.cattle.io/exclusive",
            "operator": "In",
            "values": [
              "true"
      ...
```

So for our purposes, I think we have to stick with overwriting the value of the rule because we need it to be evaluated, well, exclusively. I think what's happening here is that the code under the hood is in fact ANDing all the terms in MatchExpressions, but for anti-affinity rules it is simply negating the intersection (`~[A && B]`) which by DeMorgan's Law is actually equivalent to `~A || ~B`, hence if one of the rules evaluates to false the Pods are actually scheduled, and only if all the rules evaluate to true is the Pod not scheduled. So for anti-affinity rules, it's effectively only an AND if all the MatchExpressions evaluate to true, otherwise it's OR. I just tested this last case too and that is what I'm seeing, two rules matching Pod labels and none of the Pods are being scheduled so I'm very confident that's what's happening!